### PR TITLE
feat(`next`): add some cache primitives to data fetching

### PR DESCRIPTION
**Pull request by juliusmarminge**
Part of #3297

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

- Adds 2 new links specifically made for Next.js app router and to work nice with the Next.js cache that's available in RSC

  - 1 uses HTTP and is a small wrapper around the current httpLink and httpBatchLink
  - 1 uses the `unstable_cache` method provided by Next and invokes procedures directly

- Adds a new `revalidate` method on the client proxy that calls `revalidateTag` to invalidate the tags for a query

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: examples/.experimental/next-app-dir/src/trpc/package.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "trpc-api",
-  "typings": "server.ts",
+  "typings": "server-fetcher.ts",
   "exports": {
     ".": {
-      "types": "./server.ts",
-      "react-server": "./server.ts",
+      "types": "./server-fetcher.ts",
```
Comment (KATT) [inline examples/.experimental/next-app-dir/src/trpc/package.json:6]: `-fetcher?` 🤔

Comment (KATT) [reply examples/.experimental/next-app-dir/src/trpc/package.json:6]: Let's keep it in one example, you can see how I started splitting it up a bit more on my branch

Comment (juliusmarminge) [reply examples/.experimental/next-app-dir/src/trpc/package.json:6]: You were doing actions though?

**Review by KATT — COMMENTED** [review]
Could we add some testing to this somehow? At least something that asserts that types are working [similar to these](https://github.com/trpc/trpc/blob/7b6b270b72ab5ae13c2f2d7deb6e50899502410c/packages/next/src/app-dir/client.test.tsx#L364-L376)?

**Comment by juliusmarminge** [comment]
@KATT I reverted all the client stuff for now, there is the better implementation in the react-query-hyration repo we can use in the future. But for now, can we merge the server cache stuff and get it released?

**Review by KATT — DISMISSED** [review]

File: examples/next-prisma-websockets-starter/src/pages/api/auth/[...nextauth].ts [hunk 2]
```diff
@@ -20,10 +20,12 @@ if (useMockProvider) {
...
         if (credentials) {
+          const name =
+            typeof credentials.name === 'string' ? credentials.name : 'test';
           const user = {
-            id: credentials.name,
-            name: credentials.name,
-            email: credentials.name,
+            id: name,
+            name: name,
+            email: name,
```
Comment (juliusmarminge) [inline examples/next-prisma-websockets-starter/src/pages/api/auth/[...nextauth].ts:28]: the next-auth@5 beta im using to get a nice story for the app router seems to have changed the type here, hence the added check. could also `as any` it i guess

File: examples/.experimental/next-app-dir/src/app/client/ClientGreeting.tsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+// import { use } from 'react';
+// import { api } from 'trpc-api';
+
+export function ClientGreeting() {
+  // const greeting = use(api.greeting.query({ text: 'from client' }));
+
+  // return <>{greeting}</>;
+
+  return <>not implemented</>;
+}
```
Comment (juliusmarminge) [inline examples/.experimental/next-app-dir/src/app/client/ClientGreeting.tsx:12]: Wrote about this here; https://github.com/vercel/next.js/issues/51031

But our previous workaround with storing the promise as a ref etc seems to not be supported anymore...

Leaving this for now and we'll take this back up when doing the RQ stuff?

**Comment by juliusmarminge** [comment]
Blocked by https://github.com/blitz-js/superjson/issues/242

File: examples/.experimental/next-app-dir/src/app/api/auth/[...nextauth]/route.ts
Comment (juliusmarminge) [inline examples/.experimental/next-app-dir/src/app/api/auth/[...nextauth]/route.ts:1]: i added auth to make sure the headers and such is working correctly and that the session can be retrieved in trpc

File: examples/.experimental/next-app-dir/src/app/layout.tsx [hunk 4]
```diff
@@ -20,8 +22,19 @@ export default function RootLayout({
           style={{
```
Comment (juliusmarminge) [inline examples/.experimental/next-app-dir/src/app/layout.tsx:22]: might have to add tailwind here soon cause these big style blocks are annoying 😓

Comment (KATT) [reply examples/.experimental/next-app-dir/src/app/layout.tsx:22]: do it if you wna!

File: examples/.experimental/next-app-dir/src/app/post-example/page.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  // create the post
+  await api.createPost.mutate({
+    title: fd.get('title') as string,
+    content: fd.get('content') as string,
+  });
+
+  // revalidate the latest post
+  await api.getLatestPost.revalidate();
+}
```
Comment (KATT) [inline examples/.experimental/next-app-dir/src/app/post-example/page.tsx:14]: Could we do 

```ts
await api.createPost.mutate(fd);
```

?

**Review by KATT — APPROVED** [review]

File: examples/.experimental/next-app-dir/src/app/post-example/page.tsx (hunk 5 above)
Comment (juliusmarminge) [reply examples/.experimental/next-app-dir/src/app/post-example/page.tsx:14]: i don't think so - we do the normalization for server actions but i don't think we do it for mutations?

i can look into it in #4567 👍🏼

**Review by KATT — APPROVED** [review]

**Pull request merged**